libraries/networking/src/udt/Connection.cpp

```diff
@@ -0,0 +1,600 @@
+        
+        QObject::connect(_sendQueue.get(), &SendQueue::packetSent, this, &Connection::packetSent);
+        QObject::connect(_sendQueue.get(), &SendQueue::packetSent, this, &Connection::recordSentPackets);
+        QObject::connect(_sendQueue.get(), &SendQueue::packetRetransmitted, this, &Connection::recordRetransmission);
+        
+        // set defaults on the send queue from our congestion control object
+        _sendQueue->setPacketSendPeriod(_congestionControl->_packetSendPeriod);
+        _sendQueue->setFlowWindowSize(std::min(_flowWindowSize, (int) _congestionControl->_congestionWindowSize));
+    }
+    
+    return *_sendQueue;
+}
+
+void Connection::sendReliablePacket(unique_ptr<Packet> packet) {
+    Q_ASSERT_X(packet->isReliable(), "Connection::send", "Trying to send an unreliable packet reliably.");
+    getSendQueue().queuePacket(move(packet));
+}
+
+void Connection::sync() {
+    if (_hasReceivedFirstPacket) {
+        // reset the number of light ACKs or non SYN ACKs during this sync interval
+        _lightACKsDuringSYN = 1;
+        _acksDuringSYN = 1;
+        
+        // we send out a periodic ACK every rate control interval
+        sendACK();
+        
+        // check if we need to re-transmit a loss list
+        // we do this if it has been longer than the current nakInterval since we last sent
+        auto now = high_resolution_clock::now();
+        
+        if (duration_cast<microseconds>(now - _lastNAKTime).count() >= _nakInterval) {
+            // Send a timeout NAK packet
+            sendTimeoutNAK();
+        }
+    }
+}
+
+void Connection::recordSentPackets(int dataSize, int payloadSize) {
+    _stats.recordSentPackets(payloadSize, dataSize);
+}
+
+void Connection::recordRetransmission() {
+    _stats.record(ConnectionStats::Stats::Retransmission);
+}
+
+void Connection::sendACK(bool wasCausedBySyncTimeout) {
+    static high_resolution_clock::time_point lastACKSendTime;
+    auto currentTime = high_resolution_clock::now();
+    
+    SequenceNumber nextACKNumber = nextACK();
+    Q_ASSERT_X(nextACKNumber >= _lastSentACK, "Connection::sendACK", "Sending lower ACK, something is wrong");
+    
+    if (nextACKNumber == _lastSentACK) {
+        // We already sent this ACK, but check if we should re-send it.
+        if (nextACKNumber < _lastReceivedAcknowledgedACK) {
+            // we already got an ACK2 for this ACK we would be sending, don't bother
+            return;
+        }
+        
+        // We will re-send if it has been more than the estimated timeout since the last ACK
+        microseconds sinceLastACK = duration_cast<microseconds>(currentTime - lastACKSendTime);
+        
+        if (sinceLastACK.count() < estimatedTimeout()) {
+            return;
+        }
+    }
+    // we have received new packets since the last sent ACK
+    
+    // update the last sent ACK
+    _lastSentACK = nextACKNumber;
+    
+    // remove the ACKed packets from the receive queue
+    // TODO?
```

Comment:
we can probably remove this comment - will only come into play once we go to add ordering